[dnf5] Vars source/priority support, microdnf "--setvars" command line arg support

jrohel opened the pull request:
- Vars source/priority support
    - each variable remembers its source
    - sources have priority: DEFAULT, AUTO (eg from "detect_arch"), VARSDIR, PLUGIN, ENVIRONMENT, COMMANDLINE, RUNTIME
    - a lower priority source cannot overwrite a value from a higher priority source
    
- microdnf "--setvars" command line argument support
  The "--setvar" command line argument allows to set a new/existing libdnf variable.
  Format: `--setvar=<name>=<value>`
  Example: `--setvar=releasever=24`

jrohel committed:
Vars source/priority support

- each variable remembers its source
- sources have priority: DEFAULT, AUTO (eg from "detect_arch"),
  VARSDIR, PLUGIN, ENVIRONMENT, COMMANDLINE, RUNTIME
- a lower priority source cannot overwrite a value from a higher
  priority source

jrohel committed:
[microdnf] Support "--setvar" command line argument

The "--setvar" command line argument allows to set a new/existing libdnf
variable.
Format: --setvar=<name>=<value>
Example: --setvar=releasever=24

dmach commented on include/libdnf/conf/vars.hpp:
```diff
@@ -34,19 +34,43 @@ namespace libdnf {
 /// of directories.
 struct Vars {
 public:
+    enum class Priority { DEFAULT = 10, AUTO = 20, FILE = 30, ENVIRONMENT = 40, COMMANDLINE = 50, RUNTIME = 60 };
```
Replace `FILE` with `CONFIG`? It would probably make a better sense from a user perspective.
I looked at Option::Priority and I'm wondering if `EMPTY` and `PLUGIN` wouldn't make sense in some cases.

jrohel replied:
I was thinking about `EMPTY` variables. We have a fixed set of configuration options. But the variables behave differently. At the beginning we have only three variables `arch`,` basearch`, `releasever`. The user can add more variables. Instead of creating an `EMPTY` variable, the user does not create the variable at all. Do we want to set an existing variable to `EMPTY`? Isn't it better to introduce variable removal support?

`FILE` <-> `CONFIG` and `PLUGIN`. For discussion.

j-mracek replied:
PLUGIN priority is also available for config but it is unused. With variables I would prefer that plugin that want to modify variables will create a file in `/etc/dnf/vars` and the best would be when the file will be in RPM. Therefore I would prefer to not add `PLUGIN` priority at least from beginning.
With `FILE` <-> `CONFIG` I thing that the source is more like `VARS_DIRS` (It is a list option with "/etc/yum/vars", "/etc/dnf/vars")

j-mracek commented on include/libdnf/conf/vars.hpp:
```diff
@@ -34,19 +34,43 @@ namespace libdnf {
+    /// @brief Get value of particular variable.
+    ///
+    /// @param name Name of the variable
+    const std::string & get_value(const std::string & name) const { return variables.at(name).value; }
```
Am I correct that it raises an exception when variable is not there?

j-mracek replied:
Please could you mention that it raises an exception? The question is how we suppose to use it? When I have a map I prefer to use `find()` and then from iterator value is not `.end()`. Also I can use `.at()` but it requires handling of exceptions. In our API we have `contains` and `get_value`. Using at beginning `contains` and then `get_value` is not efficient. Do we suggest to use `get_value()` with `catch` as the best approach?

j-mracek replied:
The comment is also valid for `get()`

j-mracek commented on dnfdaemon-server/services/rpm/rpm.cpp:
```diff
@@ -225,7 +225,7 @@ libdnf::transaction::TransactionWeakPtr new_db_transaction(libdnf::Base * base,
         transaction->set_comment(comment);
     }
-    transaction->set_releasever(base->get_vars().get_values().at("releasever"));
+    transaction->set_releasever(base->get_vars().get_value("releasever"));
```
Somehow I think that the code requires that the releasever is always available or detected. I think that this is an incorrect assumption. `releasever` is an optional and there are user cases where releasever is not needed (e.g. all repositories are defined without $releasever). The problem is not in PR but already in upstream.

j-mracek commented on microdnf/main.cpp:
```diff
@@ -71,15 +72,13 @@ static bool parse_args(Context & ctx, int argc, char * argv[]) {
         auto val = strchr(value+1, '=');
         if (!val) {
-            throw std::runtime_error(std::string("setopt: Badly formated argument value") + value);
+            throw std::runtime_error(fmt::format("setopt: Badly formated argument value \"{}\"", value));
```
What about translation of the error?

j-mracek replied:
We should also set a standard to highlight the value in message: `'{}'`, `\"{}\"`, ...

jrohel replied:
@j-mracek 
- Yes translation. Good point. But this is microdnf not libdnf. I need to prepare translation support in microdnf. Probably another commit.
- We chose double quotes at the meeting.

j-mracek commented on libdnf/conf/vars.cpp:
```diff
@@ -204,21 +205,53 @@ std::string Vars::substitute(const std::string & text) const {
-    values["releasever"] = detect_release(installroot);
+    auto it = variables.find("arch");
+    if (it == variables.end()) {
+        variables.insert({"arch", {detect_arch(), Priority::AUTO}});
```
I know that it is not a problem of PR, but I am not sure whether detect_arch() is correctly implemented. It uses rpm for detection, that sets char **. Then it converts it into std::string without check for `null_ptr`. In case when the detection failed it can throw an exception and not conversion problem with std::string constructor.

j-mracek commented on libdnf/conf/vars.cpp:
```diff
@@ -204,21 +205,53 @@ std::string Vars::substitute(const std::string & text) const {
+
+    it = variables.find("basearch");
+    if (it == variables.end()) {
+        variables.insert({"basearch", {get_base_arch(variables["arch"].value.c_str()), Priority::AUTO}});
```
`get_base_arch()` can return nullptr, therefore it can result in crash. What about to test it and throw an exception when unsupported `basearch` appears.

j-mracek commented on libdnf/conf/vars.cpp:
```diff
@@ -204,21 +205,53 @@ std::string Vars::substitute(const std::string & text) const {
+
+    it = variables.find("releasever");
+    if (it == variables.end()) {
+        variables.insert({"releasever", {detect_release(installroot), Priority::AUTO}});
```
I am really sorry but I looked at `detect_release()` and I am not happy with it. It returns an empty string when detection fails. The empty string is set in vars therefore $releasever in URL is replaced by empty string. Then it reports that URL of repository is not valid, but it is difficult to detect that the problem is in detection of releasever and then in substitution. Again this not a problem of PR.

jrohel committed:
Improve detect_arch() and detect_release()

detect_arch():
Can "rpmGetArchInfo" return nullptr as name?
detect_arch() now returns original "const char *" instead of "std::string".

detect_release():
Returns "std::optional<std::string>" instead of "std::string".
To distinguish between "" and not found.
Only non-empty providers "evr" is used.

Fixes "Vars::detect_vars()" for nullptr values.

j-mracek commented:
LGTM
